### PR TITLE
Welcome to 2024

### PR DESCRIPTION
**Pull request by JustArchi**

### CODE REVIEW

File: ItemDispenser/ItemDispenser.cs [hunk 1]
```diff
@@ -1,101 +1,106 @@
...
-				bot.ArchiLogger.LogGenericError("Item Dispenser configuration is wrong!");
-				BotSettings.AddOrUpdate(bot, [], (k, v) => []);
-			}
-			return Task.CompletedTask;
+		byte? holdDuration = await bot.GetTradeHoldDuration(tradeOffer.OtherSteamID64, tradeOffer.TradeOfferID).ConfigureAwait(false);
+
+		// If user has a trade hold, we add extra logic
+		// If trade hold duration exceeds our max, or user asks for cards with short lifespan, reject the trade
+		// If we can't get trade hold duration, ignore
+		switch (holdDuration) {
```
Comment (Rudokhvist) [inline ItemDispenser/ItemDispenser.cs:92]: There is no reason to use `switch` with a single `case`. Please replace it with a single `if`.

> Warning	IDE0010	Populate switch

File: ItemDispenser/ItemDispenser.cs [hunk 2]
```diff
@@ -1,101 +1,106 @@
...
-		}
+			BotSettings[bot] = botSettings.ToFrozenDictionary(static kv => kv.Key, static kv => kv.Value.ToFrozenSet());
+		} catch (Exception e) {
+			bot.ArchiLogger.LogGenericException(e);
+			bot.ArchiLogger.LogGenericError("Item Dispenser configuration is wrong!");
 
-		public Task OnLoaded() {
-			ASF.ArchiLogger.LogGenericInfo("Item Dispenser Plugin by Rudokhvist, powered by ginger cats");
-			return Task.CompletedTask;
+			BotSettings.TryRemove(bot, out _);
```
Comment (Rudokhvist) [inline ItemDispenser/ItemDispenser.cs:67]: > Warning	IDE0058	Expression value is never used

File: ItemDispenser/ItemDispenser.cs [hunk 3]
```diff
@@ -1,101 +1,106 @@
...
+
+			return Task.CompletedTask;
+		}
 
-			// If user has a trade hold, we add extra logic
-			if (holdDuration.Value > 0) {
-				// If trade hold duration exceeds our max, or user asks for cards with short lifespan, reject the trade
-				if ((holdDuration.Value > (ASF.GlobalConfig?.MaxTradeHoldDuration ?? 0)) || tradeOffer.ItemsToGiveReadOnly.Any(item => ((item.Type == Asset.EType.FoilTradingCard) || (item.Type == Asset.EType.TradingCard)) && CardsFarmer.SalesBlacklist.Contains(item.RealAppID))) {
-					return false;
+		Dictionary<(uint AppID, ulong ContextID), HashSet<Asset.EType>> botSettings = new();
```
Comment (Rudokhvist) [inline ItemDispenser/ItemDispenser.cs:44]: > Warning	IDE0028	Collection initialization can be simplified

File: ItemDispenser/ItemDispenser.cs [hunk 4]
```diff
@@ -1,101 +1,106 @@
...
 
-			foreach (Asset item in tradeOffer.ItemsToGiveReadOnly) {
-				if (!ItemsToDispense.Any(sample =>
-									   (sample.AppID == item.AppID) &&
-									   (sample.ContextID == item.ContextID) &&
-									   (sample.Types.Count <= 0 || sample.Types.Any(type => type == item.Type))
-										)) {
-					return false;
+				if (!botSettings.TryGetValue(key, out HashSet<Asset.EType>? types)) {
+					types = new HashSet<Asset.EType>();
```
Comment (Rudokhvist) [inline ItemDispenser/ItemDispenser.cs:55]: > Warning	IDE0028	Collection initialization can be simplified

**Comment by JustArchi** [comment]
I'm not fixing VS stupidity, every single warning you've linked @Rudokhvist is a false positive - there is a reason why those start with `IDE` and are not part of msbuild, like all other ones (i.e. `CS`). Take it or leave it :shrug:

File: ItemDispenser/ItemDispenser.cs [hunk 5]
```diff
@@ -1,101 +1,106 @@
...
+	public Task OnBotInitModules(Bot bot, IReadOnlyDictionary<string, JsonElement>? additionalConfigProperties) {
+		ArgumentNullException.ThrowIfNull(bot);
 
-			//If we receiveing something in return, and donations is not accepted - ignore.
-			if (tradeOffer.ItemsToReceiveReadOnly.Count > 0 && !bot.BotConfig.TradingPreferences.HasFlag(BotConfig.ETradingPreferences.AcceptDonations)) {
-				return false;
-			}
-			byte? holdDuration = await bot.GetTradeHoldDuration(tradeOffer.OtherSteamID64, tradeOffer.TradeOfferID).ConfigureAwait(false);
+		if (additionalConfigProperties == null) {
+			BotSettings.TryRemove(bot, out _);
```
[the remark this answers is not in the document]
Comment (JustArchi) [reply ItemDispenser/ItemDispenser.cs:33]: This is precisely why it's a discard `_` variable :man_facepalming:

File: ItemDispenser/ItemDispenser.cs [hunk 6]
```diff
@@ -1,101 +1,106 @@
...
 
-			if (!holdDuration.HasValue) {
-				// If we can't get trade hold duration, ignore
-				return false;
-			}
+			return Task.CompletedTask;
+		}
+
+		if (!additionalConfigProperties.TryGetValue("Rudokhvist.DispenseItems", out JsonElement jsonElement)) {
+			BotSettings.TryRemove(bot, out _);
```
[the remark this answers is not in the document]
Comment (JustArchi) [reply ItemDispenser/ItemDispenser.cs:39]: This is precisely why it's a discard `_` variable :man_facepalming:

File: ItemDispenser/ItemDispenser.cs (hunk 2 above)
Comment (JustArchi) [reply ItemDispenser/ItemDispenser.cs:67]: This is precisely why it's a discard `_` variable :man_facepalming:

File: ItemDispenser/ItemDispenser.cs [hunk 7]
```diff
@@ -1,101 +1,106 @@
...
-using ArchiSteamFarm.Helpers.Json;
 
-namespace ItemDispenser {
-	[Export(typeof(IPlugin))]
-	public class ItemDispenser : IBotTradeOffer, IBotModules {
+namespace ItemDispenser;
 
-		private readonly ConcurrentDictionary<Bot, ConcurrentHashSet<DispenseItem>> BotSettings = new();
+[Export(typeof(IPlugin))]
+[SuppressMessage("ReSharper", "UnusedType.Global")]
```
[the remark this answers is not in the document]
Comment (JustArchi) [reply ItemDispenser/ItemDispenser.cs:21]: It's necessary for Rider, not surprised VS can't see anything beyond its own nose.

[the remark this answers is not in the document]
Comment (Rudokhvist) [reply ItemDispenser/ItemDispenser.cs:21]: When you will buy me a Rider license - I will start to care about it. For now it shows a notice in VS, and that's all I care about.

File: ItemDispenser/ItemDispenser.cs (hunk 2 above)
Comment (Rudokhvist) [reply ItemDispenser/ItemDispenser.cs:67]: It's a warning about another discard. TryRemove returns a value, and it has to be explicitly discaded.

File: ItemDispenser/ItemDispenser.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Rudokhvist) [reply ItemDispenser/ItemDispenser.cs:39]: It's a warning about another discard. TryRemove returns a value, and it has to be explicitly discaded.

File: ItemDispenser/ItemDispenser.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rudokhvist) [reply ItemDispenser/ItemDispenser.cs:33]: It's a warning about another discard 🤦‍♂️. TryRemove returns a value, and it has to be explicitly discarded.

File: ItemDispenser/ItemDispenser.cs (hunk 1 above)
Comment (Rudokhvist) [reply ItemDispenser/ItemDispenser.cs:92]: The more cases switch has - the more slower it works than if. Proof: https://cc.davelozinski.com/c-sharp/what-is-the-fastest-conditional-statement
So you have two cases for the same ONE branch - it's literally worst thing you can do,

**Pull request closed without merging**